packages/ui/src/components/ChatList.tsx

```diff
@@ -242,164 +150,273 @@ export function ChatList({
+            name="messages"
+            activeTab={activeTab}
+            onTabPress={onPressTab}
+          >
+            <Tabs.Title active={activeTab === 'messages'}>Messages</Tabs.Title>
+          </Tabs.Tab>
+        </Tabs>
+      </YStack>
+    </Animated.View>
+  );
+}
+
+function useFilteredChats({
+  pinned,
+  unpinned,
+  pending,
+  searchQuery,
+  activeTab,
+}: {
+  pinned: Chat[];
+  unpinned: Chat[];
+  pending: Chat[];
+  searchQuery: string;
+  activeTab: TabName;
+}) {
+  const performSearch = useChatSearch({ pinned, unpinned });
+  const debouncedQuery = useDebouncedValue(searchQuery, 200);
+  const searchResults = useMemo(
+    () => performSearch(debouncedQuery),
+    [debouncedQuery, performSearch]
+  );
+
+  return useMemo(() => {
+    const isSearching = searchQuery.trim() !== '';
+    if (!isSearching) {
+      const pinnedSection = {
+        title: 'Pinned',
+        data: filterChats(pinned, activeTab),
+      };
+      const allSection = {
+        title: 'All',
+        data: [...pending, ...filterChats(unpinned, activeTab)],
+      };
+      return pinnedSection.data.length
+        ? [pinnedSection, allSection]
+        : [allSection];
     } else {
-      filterVisible.value = false;
-      scrollY.value = withSpring(0);
+      return [{ title: 'Search', data: filterChats(searchResults, activeTab) }];
     }
-  }, [showFilters, filterVisible, scrollY, FILTER_HEIGHT]);
+  }, [activeTab, pending, searchQuery, searchResults, unpinned, pinned]);
+}
 
-  const getChannelKey = useCallback((item: unknown) => {
-    const chatItem = item as Chat;
+function filterChats(chats: Chat[], activeTab: TabName) {
+  if (activeTab === 'all') return chats;
+  return chats.filter((chat) => {
+    const isGroupChannel = logic.isGroupChannelId(chat.id);
+    return activeTab === 'groups' ? isGroupChannel : !isGroupChannel;
+  });
+}
 
-    if (!chatItem || typeof chatItem !== 'object' || !chatItem.id) {
-      return 'invalid-item';
-    }
+function useChatSearch({
+  pinned,
+  unpinned,
+}: {
+  pinned: Chat[];
+  unpinned: Chat[];
+}) {
+  const fuse = useMemo(() => {
+    const allData = [...pinned, ...unpinned];
+    return new Fuse(allData, {
+      keys: [
+        'id',
+        'group.title',
+        'contact.nickname',
+        'members.contact.nickname',
+        'members.contact.id',
+      ],
+      threshold: 0.3,
+    });
+  }, [pinned, unpinned]);
 
-    if (logic.isGroup(chatItem)) {
-      return chatItem.id;
-    }
-    return `${chatItem.id}-${chatItem.pin?.itemId ?? ''}`;
-  }, []);
+  const performSearch = useCallback(
+    (query: string) => {
+      return fuse.search(query).map((result) => result.item);
+    },
+    [fuse]
+  );
+
+  return performSearch;
+}
+
+function useDebouncedValue<T>(input: T, delay: number) {
+  const [value, setValue] = useState<T>(input);
+  const debouncedSetValue = useMemo(
+    () => debounce(setValue, delay, { leading: true }),
+    [delay]
+  );
+  useLayoutEffect(() => {
```

Comment:
why `useLayoutEffect` here? I think of `useLayoutEffect` as like "make this change blocking a repaint because it will likely change the outcome of the repaint" – I think `debounce` will always call `setValue` asynchronously, which means we'll never get a change to `value` synchronously before the repaint; so a `useEffect` should work as well and prevent adding work that blocks a render. (i'm not convinced that i understand useLayoutEffect, lmk if that's wrong)